### PR TITLE
feat: Updates to the Ray blueprint

### PR DESCRIPTION
**Pull request by askulkarni2**
### What does this PR do?

🛑 Please open an issue first to discuss any significant work and flesh out details/direction - we would hate for your time to be wasted.
Consult the [CONTRIBUTING](https://github.com/awslabs/data-on-eks/blob/main/CONTRIBUTING.md#contributing-via-pull-requests) guide for submitting pull-requests.

### Motivation

The current example for Ray under AI/ML is outdated and uses some of the legacy code. This PR updates the example to incorporate the new helm charts for kuberay-operator. The example also takes into account the upcoming changes in the v5 of EKS Blueprints and hence uses upstream EKS module for cluster and node group creation.

### More

- [x] Yes, I have tested the PR using my local account setup (Provide any test evidence report under Additional Notes)
- [ ] Yes, I have added a new example under [examples](https://github.com/awslabs/data-on-eks/tree/main/examples) to support my PR
- [ ] Yes, I have created another PR to add a new Add-on for [Terraform EKS Blueprints](https://github.com/aws-ia/terraform-aws-eks-blueprints) repo (if applicable)
- [x] Yes, I have updated the `website/docs` or `website/blog` section for this feature
- [x] Yes, I ran `pre-commit run -a` with this PR. Link for installing [pre-commit](https://pre-commit.com/) locally

### For Moderators

- [ ] E2E Test successfully complete before merge?

### CODE REVIEW

File: ai-ml/ray/terraform/variables.tf [hunk 1]
```diff
@@ -0,0 +1,17 @@
+variable "region" {
+  description = "AWS Region"
+  type        = string
+  default     = "us-east-1"
```
Comment (vara-bonthu) [inline ai-ml/ray/terraform/variables.tf:4]: Are we not using `us-west-2` as default?

File: ai-ml/ray/terraform/examples/pytorch/main.tf [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    }
+  }
+}
+
+data "aws_eks_cluster" "this" {
+  name = local.eks_cluster
+}
+
+locals {
+  region      = "us-east-1"
```
Comment (vara-bonthu) [inline ai-ml/ray/terraform/examples/pytorch/main.tf:36]: region us-west-2?

File: ai-ml/ray/terraform/examples/pytorch/main.tf [hunk 3]
```diff
@@ -0,0 +1,122 @@
+provider "aws" {
+  region = local.region
+}
+
+provider "kubernetes" {
```
Comment (vara-bonthu) [inline ai-ml/ray/terraform/examples/pytorch/main.tf:5]: This approach comes with pre-requisites and I don't see this approach solved the original problem. Data source is working fine and no issues so far. 

```tf
provider "kubernetes" {
  host                   = module.eks_blueprints.eks_cluster_endpoint
  cluster_ca_certificate = base64decode(module.eks_blueprints.eks_cluster_certificate_authority_data)
  token                  = data.aws_eks_cluster_auth.this.token
}
```

File: ai-ml/ray/terraform/examples/pytorch/main.tf [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  eks_cluster_name = local.eks_cluster
+
+  helm_values = [
+    yamlencode({
+      image = {
+        repository = "rayproject/ray-ml"
+        # This is a different version than the xgboost version
+        tag        = "2.3.0"
+        pullPolicy = "IfNotPresent"
+      }
```
Comment (vara-bonthu) [inline ai-ml/ray/terraform/examples/pytorch/main.tf:55]: In my opinion, YAML is a big advantage for Kubernetes as it keeps things simple. It's much easier to work with, especially when editing configuration files. Many customers find it easier to update the YAML file directly with new configuration parameters, rather than having to modify Terraform code or learn how to convert YAML to Terraform. This simplicity and ease of use is what makes YAML a popular choice for Kubernetes configurations.

File: ai-ml/ray/terraform/examples/xgboost/main.tf [hunk 5]
```diff
@@ -0,0 +1,115 @@
+provider "aws" {
+  region = local.region
+}
+
+provider "kubernetes" {
```
Comment (vara-bonthu) [inline ai-ml/ray/terraform/examples/xgboost/main.tf:5]: See above comment on provider

File: ai-ml/ray/terraform/main.tf [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+#---------------------------------------------------------------
+
+provider "aws" {
+  region = local.region
+}
+
+# Used for Karpenter Helm chart
+provider "aws" {
+  region = "us-east-1"
+  alias  = "virginia"
```
Comment (vara-bonthu) [inline ai-ml/ray/terraform/main.tf:12]: May be we can call this with meaningful name like something like `ecr_public_region` ?

File: website/docs/ai-ml/ray.md [hunk 7]
```diff
@@ -6,317 +6,337 @@ sidebar_label: Ray on EKS
...
-Finally, destroy any additional resources that are not in the above modules
+Delete everything else.
 
 ```bash
 terraform destroy -auto-approve
 ```
+
+### Future Work
+
+We want to enhance this blueprint to add functionality that makes this pattern more enterprise ready. To that end in the near future we will be adding enhancements around cost management using kubecost, better observability dashboards using OTEl, Amazon Managed Prometheus and Grafana, better security and data governance using OPA/Gatekeeper, IRSA, etc. Please let us know via [issues](https://github.com/awslabs/data-on-eks/issues) to see what you would like to see us build in this blueprint.
```
Comment (vara-bonthu) [inline website/docs/ai-ml/ray.md:342]: Lets add this as a Note and move to the top of the doc

:::info

As part of our ongoing efforts to make this blueprint more enterprise-ready, we are actively working on adding several key functionalities. This includes cost management with Kubecost, advanced observability with OTEL, Amazon Managed Prometheus, and Grafana, as well as improved security and data governance using tools such as OPA/Gatekeeper and IRSA. If you have specific requirements or suggestions for this blueprint, please feel free to open an issue on our[ GitHub repository](https://github.com/awslabs/data-on-eks/issues).

:::

**Review by vara-bonthu — COMMENTED** [review]
Thanks for Updated Ray Blueprint 👍🏼 It looks great. 
I left few monitor comments. I didn't get a chance to deploy this blueprint but I will do as soon as i get some free time.

File: ai-ml/ray/terraform/main.tf [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+  eks_cluster_version  = module.eks.cluster_version
+
+  # Wait on the node group(s) before provisioning addons
+  data_plane_wait_arn = join(",", [for group in module.eks.eks_managed_node_groups : group.node_group_arn])
+
+  enable_aws_cloudwatch_metrics = true
+  aws_cloudwatch_metrics_helm_config = {
+    version = "0.0.8"
+  }
+
```
[the remark this answers is not in the document]
Comment (vara-bonthu) [reply ai-ml/ray/terraform/main.tf:347]: seen the comment below

File: ai-ml/ray/terraform/main.tf [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+  secondary_cidr_blocks = [local.secondary_vpc_cidr]
+
+  azs = local.azs
+  private_subnets = concat(
+    [for k, v in local.azs : cidrsubnet(local.vpc_cidr, 4, k)],
+    [for k, v in local.azs : cidrsubnet(local.secondary_vpc_cidr, 4, k)]
+  )
+  public_subnets = [for k, v in local.azs : cidrsubnet(local.vpc_cidr, 8, k + 48)]
+  # Control Plane Subnets
+  intra_subnets = [for k, v in local.azs : cidrsubnet(local.vpc_cidr, 8, k + 52)]
```
Comment (vara-bonthu) [inline ai-ml/ray/terraform/main.tf:106]: Curious about `intra_subnets`. Any documentation around this and why is it recommended? Or Add some comments just above that.

File: ai-ml/ray/terraform/main.tf [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+    Blueprint  = local.name
+    GithubRepo = "github.com/awslabs/data-on-eks"
+  }
+}
+
+#---------------------------------------------------------------
+# VPC
+#---------------------------------------------------------------
+
+module "vpc" {
```
Comment (vara-bonthu) [inline ai-ml/ray/terraform/main.tf:90]: Move the VPC resources to `vpc.tf` file and you can leave the rest of the resources in this file.

File: ai-ml/ray/terraform/main.tf (hunk 6 above)
Comment (askulkarni2) [reply ai-ml/ray/terraform/main.tf:12]: fixed

File: ai-ml/ray/terraform/variables.tf (hunk 1 above)
Comment (askulkarni2) [reply ai-ml/ray/terraform/variables.tf:4]: fixed

File: ai-ml/ray/terraform/main.tf [hunk 11]
```diff
@@ -0,0 +1,318 @@
...
+          AWS_VPC_K8S_CNI_CUSTOM_NETWORK_CFG = "true"
+          ENI_CONFIG_LABEL_DEF               = "topology.kubernetes.io/zone"
+
+          # Reference docs https://docs.aws.amazon.com/eks/latest/userguide/cni-increase-ip-addresses.html
+          ENABLE_PREFIX_DELEGATION = "true"
+          WARM_PREFIX_TARGET       = "1"
+        }
+      })
+    }
+  }
```
Comment (vara-bonthu) [inline ai-ml/ray/terraform/main.tf:145]: @askulkarni2 Let's move this to Kubernetes add-ons module in the next revision.

**Review by vara-bonthu — APPROVED** [review]
LGTM 👍🏼
